### PR TITLE
[tune] deflake pbt.

### PR DESCRIPTION
**Pull request by xwjiang2010**
## Why are these changes needed?
We use `trial.checkpoint` to restore a perturbed trial. Currently trial.checkpoint is looking at both in-memory and persistent checkpoints to find the most recent one. The definition of "the most recent one" is based on iteration. This may no longer be a valid assumption in PBT case, considering `trial_low_quantile` may have an iter=2_persistent_checkpoint as well as a iter=1_in_memory_checkpoint (perturbed from `trial_upper_quantile`). 

BTW, I think there are a few other issues to clean up for PBT case, depending on how important we think PBT is. For example,
- multiple times of checkpoint for the same trial at the same stage (trial_upper.save(MEMORY), trial_upper.save(PERSISTENT), trial_upper.pause() (which includes a trial_upper.save(MEMORY)) )
- the potential data inconsistency between PBT.trial_state and real trial state. Ideally PBT should not keep its copy of trial state.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/tune/checkpoint_manager.py [hunk 1]
```diff
@@ -114,6 +118,11 @@ def replace_newest_memory_checkpoint(self, new_checkpoint):
         self._newest_memory_checkpoint = new_checkpoint
 
     def on_checkpoint(self, checkpoint):
+        self._cur_order += 1
+        checkpoint.order = self._cur_order
+        self.on_checkpoint_internal(checkpoint)
+
+    def on_checkpoint_internal(self, checkpoint):
```
Comment (krfricke) [inline python/ray/tune/checkpoint_manager.py:125]: Do we need the `on_checkpoint_internal` function here? Seems like we never call it anywhere else

**Review by krfricke — APPROVED** [review]
Great discovery, thanks!

**Comment by krfricke** [comment]
To fix the build errors, merge current master

File: python/ray/tune/checkpoint_manager.py (hunk 1 above)
Comment (xwjiang2010) [reply python/ray/tune/checkpoint_manager.py:125]: oh right, making it private

Comment (matthewdeng) [reply python/ray/tune/checkpoint_manager.py]: What's the purpose of splitting out `on_checkpoint_internal` into a separate method altogether? 

What logic should be contained in the external method vs. the internal method?

Comment (krfricke) [reply python/ray/tune/checkpoint_manager.py]: Yep, would also opt to remove the function completely and just keep `on_checkpoint`

Comment (xwjiang2010) [reply python/ray/tune/checkpoint_manager.py]: sure. combining the two.

**Review by krfricke — APPROVED** [review]
Awesome, merging after tests pass

**Pull request merged**